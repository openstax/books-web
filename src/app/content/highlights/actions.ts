import { createStandardAction } from 'typesafe-actions';

export const enableHighlighting = createStandardAction('FeatureFlag/enableHighlighting')<void>();
