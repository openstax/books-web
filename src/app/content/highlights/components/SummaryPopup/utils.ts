import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const createHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
  const data = { target: JSON.stringify({ id, type: 'highlight' }) };
  const target = queryString.stringify(data) + `#${elementId}`;
  return `${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`;
};
