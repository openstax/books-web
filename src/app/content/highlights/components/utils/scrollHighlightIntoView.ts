import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import { scrollIntoView } from '../../../../domUtils';

export default (highlight: Highlight, element?: HTMLElement | null) => {
  const firstElement = highlight.elements[0] as HTMLElement;
  const lastElement = highlight.elements[highlight.elements.length - 1] as HTMLElement;
  const otherElements = [lastElement];
  if (element) {
    otherElements.push(element);
  }
  scrollIntoView(firstElement, otherElements);
};
