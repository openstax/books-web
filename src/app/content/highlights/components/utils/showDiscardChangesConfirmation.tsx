import { MiddlewareAPI } from '../../../../types';
import { showConfirmationModal } from '../../../actions';
import { setForceScrollToHiglight } from '../../actions';

export default (
  dispatch: MiddlewareAPI['dispatch']): Promise<boolean> => new Promise((resolve) => {
    dispatch(showConfirmationModal({
      options: {
        bodyi18nKey: 'i18n:discard:body',
        callback: (confirmed: boolean) => {
          resolve(confirmed);
          if (!confirmed) {
            dispatch(setForceScrollToHiglight(true));
          }
        },
        cancelButtoni18nKey: 'i18n:discard:button:cancel',
        headingi18nKey: 'i18n:discard:heading',
        okButtoni18nKey: 'i18n:discard:button:discard',
      },
    }));
  });
