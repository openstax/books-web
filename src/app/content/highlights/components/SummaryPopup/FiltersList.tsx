import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import styled from 'styled-components/macro';
import { PlainButton } from '../../../../components/Button';
import Times from '../../../../components/Times';
import { textStyle } from '../../../../components/Typography';
import { match, not } from '../../../../fpUtils';
import theme from '../../../../theme';
import { disablePrint } from '../../../components/utils/disablePrint';
import { splitTitleParts } from '../../../utils/archiveTreeUtils';
import { setSummaryFilters } from '../../actions';
import { highlightLocationFilters, summaryColorFilters, summaryLocationFilters } from '../../selectors';

// tslint:disable-next-line: variable-name
export const StyledPlainButton = styled(PlainButton)`
  height: 1.7rem;
  margin-right: 0.4rem;

  svg {
    height: 0.8rem;
    width: 0.8rem;
    color: ${theme.color.primary.gray.base};
  }

  ${disablePrint}
`;

// tslint:disable-next-line: variable-name
const ItemLabel = styled.span`
  ${textStyle}
  font-weight: 300;
  color: ${theme.color.primary.gray.base};
  max-width: 80px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-transform: capitalize;
  line-height: 1.5rem;

  @media print {
    max-width: max-content;
  }
`;

// tslint:disable-next-line: variable-name
const FilterListItem = styled.li`
  margin-right: 3.2rem;
  display: flex;
  align-items: center;
  overflow: visible;
  height: 4rem;
`;

interface FiltersListColorProps {
  color: HighlightColorEnum;
  onRemove: () => void;
}

// tslint:disable-next-line: variable-name
export const FiltersListColor = (props: FiltersListColorProps) => (
  <FilterListItem>
    <FormattedMessage id='i18n:highlighting:filters:remove:color' values={{filterValue: props.color}}>
      {(msg: string) => <StyledPlainButton aria-label={msg} onClick={props.onRemove}>
        <Times />
      </StyledPlainButton>}
    </FormattedMessage>

    <ItemLabel>
      <FormattedMessage id={`i18n:highlighting:colors:${props.color}`}>
        {(msg: string) => msg}
      </FormattedMessage>
    </ItemLabel>
  </FilterListItem>
);

interface FiltersListChapterProps {
  title: string;
  locationId: string;
  onRemove: () => void;
}

// tslint:disable-next-line: variable-name
export const FiltersListChapter = (props: FiltersListChapterProps) => (
  <FilterListItem>
    <FormattedMessage
      id='i18n:highlighting:filters:remove:chapter'
      values={{filterValue: splitTitleParts(props.title).join(' ')}}
    >
      {(msg: string | Element) => <StyledPlainButton aria-label={msg} onClick={props.onRemove}>
        <Times />
      </StyledPlainButton>}
    </FormattedMessage>
    <ItemLabel dangerouslySetInnerHTML={{ __html: props.title }} />
  </FilterListItem>
);

interface FiltersListProps {
  className?: string;
}

// tslint:disable-next-line: variable-name
const FiltersList = ({className}: FiltersListProps) => {
  const locationFilters = useSelector(highlightLocationFilters);
  const selectedColorFilters = useSelector(summaryColorFilters);
  const selectedLocationFilters = useSelector(summaryLocationFilters);
  const dispatch = useDispatch();

  const onRemoveChapter = (locationId: string) => {
    dispatch(setSummaryFilters({
      locationIds: [...selectedLocationFilters].filter(not(match(locationId))),
    }));
  };

  const onRemoveColor = (color: HighlightColorEnum) => {
    dispatch(setSummaryFilters({
      colors: [...selectedColorFilters].filter(not(match(color))),
    }));
  };

  return <ul className={className}>
    {Array.from(locationFilters).map(([locationId, location]) => selectedLocationFilters.has(locationId) &&
    <FiltersListChapter
      key={locationId}
      title={location.title}
      locationId={locationId}
      onRemove={() => onRemoveChapter(locationId)}
    />)}
    {[...selectedColorFilters].sort().map((color) => <FiltersListColor
      key={color}
      color={color}
      onRemove={() => onRemoveColor(color)}
    />)}
  </ul>;
};

export default styled(FiltersList)`
  ${textStyle}
  font-size: 1.4rem;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  margin: 0;
  padding: 0.4rem 0;
  list-style: none;
  overflow: visible;

  @media print {
    margin: 0;
  }
`;
