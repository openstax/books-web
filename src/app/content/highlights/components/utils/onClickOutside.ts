import { EventListenerOptions, HTMLElement, MouseEvent } from '@openstax/types/lib.dom';
import React from 'react';
import { assertDocument, assertWindow } from '../../../../utils';

const onClickOutside = (
  element: React.RefObject<HTMLElement>,
  isFocused: boolean,
  cb: (ev: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => () => {
  if (typeof document === 'undefined') {
    return () => null;
  }

  const ifOutside = (e: MouseEvent) => {
    if (!(e.target instanceof assertWindow().Element)) {
      return;
    }
    if (!element.current || element.current.contains(e.target)) {
      return;
    }
    cb(e);
  };

  if (isFocused) {
    document.addEventListener('click', ifOutside, eventOptions);
  }

  return () => assertDocument().removeEventListener('click', ifOutside, eventOptions);
};

export const useOnClickOutside = (
  element: React.RefObject<HTMLElement>,
  isEnabled: boolean,
  cb: (e: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => {
  React.useEffect(onClickOutside(element, isEnabled, cb, eventOptions), [isEnabled]);
};

export default onClickOutside;
