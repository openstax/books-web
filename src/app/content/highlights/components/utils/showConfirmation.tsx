import React from 'react';
import ReactDOM from 'react-dom';
import uuid from 'uuid/v4';
import TestContainer from '../../../../../test/TestContainer';
import { assertDocument, assertNotNull } from '../../../../utils';
import ConfirmationModal from '../ConfirmationModal';

export default async() => {
  const document = assertDocument();
  const domNode = document.createElement('div');

  domNode.id = `dialog-${uuid()}`;
  const root = assertNotNull(document.getElementById('root'), 'root element not found');
  root.insertAdjacentElement('afterend', domNode);

  const userChoice: boolean = await new Promise((resolve) => {
    const confirm = () => resolve(true);
    const deny = () => resolve(false);
    ReactDOM.render(
      <TestContainer>
        <ConfirmationModal deny={deny} confirm={confirm} />
      </TestContainer>,
      domNode
    );
  });

  ReactDOM.unmountComponentAtNode(domNode);
  assertNotNull(domNode.parentElement, 'parent element not found').removeChild(domNode);

  return userChoice;
};
