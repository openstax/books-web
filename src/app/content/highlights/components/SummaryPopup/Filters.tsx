import flow from 'lodash/fp/flow';
import React from 'react';
import { connect } from 'react-redux';
import { AppState, Dispatch } from '../../../../types';
import ChapterFilter from '../../../elements/popUp/ChapterFilter';
import Filters, { FilterDropdown } from '../../../elements/popUp/Filters';
import FiltersList from '../../../elements/popUp/FiltersList';
import PrintButton from '../../../elements/popUp/PrintButton';
import { printSummaryHighlights, setSummaryFilters } from '../../actions';
import * as select from '../../selectors';
import ColorFilter from './ColorFilter';

// tslint:disable-next-line:variable-name
export const ConnectedChapterFilter = connect(
  (state: AppState) => ({
    locationFilters: select.highlightLocationFilters(state),
    locationFiltersWithContent: select.highlightLocationFiltersWithContent(state),
    selectedLocationFilters: select.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(ChapterFilter);

// tslint:disable-next-line:variable-name
export const ConnectedFilterList = connect(
  (state: AppState) => ({
    locationFilters: select.highlightLocationFilters(state),
    selectedColorFilters: select.summaryColorFilters(state),
    selectedLocationFilters: select.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(FiltersList);

// tslint:disable-next-line:variable-name
export const ConnectedPrintButton = connect(
  (state: AppState) => ({
    disabled: select.summaryIsLoading(state),
    loading: select.summaryIsLoading(state),
    shouldFetchMore: select.hasMoreResults(state),
  }),
  (dispatch: Dispatch) => ({
    loadHighlightsAndPrint: flow(printSummaryHighlights, dispatch),
  }),
  (stateProps, dispatchProps, ownProps) => {
    const {shouldFetchMore, loadHighlightsAndPrint, ...props} = {
      ...stateProps,
      ...dispatchProps,
      ...ownProps,
    };

    return shouldFetchMore
      ? {...props, onClick: loadHighlightsAndPrint}
      : props
    ;
  }
)(PrintButton);

export default () =>
  <Filters>
    <FilterDropdown
      label='i18n:highlighting:filters:chapters'
      ariaLabelId='i18n:highlighting:filters:filter-by:aria-label'
    >
      <ConnectedChapterFilter />
    </FilterDropdown>
    <FilterDropdown
      label='i18n:highlighting:filters:colors'
      ariaLabelId='i18n:highlighting:filters:filter-by:aria-label'
    >
      <ColorFilter />
    </FilterDropdown>
    <ConnectedPrintButton />
    <ConnectedFilterList />
  </Filters>;
