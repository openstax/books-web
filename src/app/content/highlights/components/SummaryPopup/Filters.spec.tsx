import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../../test/createTestServices';
import createTestStore from '../../../../../test/createTestStore';
import { book as archiveBook, pageInChapter } from '../../../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../../../test/mocks/osWebLoader';
import MessageProvider from '../../../../MessageProvider';
import { MiddlewareAPI, Store } from '../../../../types';
import { assertDefined } from '../../../../utils';
import { receiveBook, receivePage } from '../../../actions';
import { formatBookData } from '../../../utils';
import { findArchiveTreeNode } from '../../../utils/archiveTreeUtils';
import { stripIdVersion } from '../../../utils/idUtils';
import { receiveHighlightsTotalCounts, setSummaryFilters } from '../../actions';
import { printSummaryHighlights } from '../../actions';
import Filters from './Filters';
import { FiltersListChapter, FiltersListColor, StyledPlainButton } from './FiltersList';

jest.mock('./ColorFilter', () => (props: any) => <div mock-color-filter {...props} />);
jest.mock('./ChapterFilter', () => (props: any) => <div mock-chapter-filter {...props} />);

describe('Filters', () => {
  let store: Store;
  const book = formatBookData(archiveBook, mockCmsBook);
  let helpers: ReturnType<typeof createTestServices> & MiddlewareAPI;
  let dispatch: jest.SpyInstance;
  let storeDispatch: jest.SpyInstance;

  beforeEach(() => {
    store = createTestStore();

    helpers = {
      ...createTestServices(),
      dispatch: store.dispatch,
      getState: store.getState,
    };

    dispatch = jest.spyOn(helpers, 'dispatch');
    storeDispatch = jest.spyOn(store, 'dispatch');
  });

  it('matches snapshot', () => {
    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <Filters />
      </MessageProvider>
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('updates on summary filters change', () => {
    store.dispatch(receiveBook(book));

    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <Filters />
      </MessageProvider>
    </Provider>);

    let chapterFilters = component.root.findAllByType(FiltersListChapter);
    let colorFilters = component.root.findAllByType(FiltersListColor);

    expect(chapterFilters.length).toEqual(0);
    expect(colorFilters.length).toEqual(5);

    const pageId = stripIdVersion(book.tree.contents[0].id);
    const chapterId = stripIdVersion(book.tree.contents[1].id);
    renderer.act(() => {
      store.dispatch(setSummaryFilters({
        colors: [HighlightColorEnum.Blue, HighlightColorEnum.Yellow],
        locationIds: [pageId, chapterId],
      }));
    });

    chapterFilters = component.root.findAllByType(FiltersListChapter);
    colorFilters = component.root.findAllByType(FiltersListColor);

    expect(chapterFilters.length).toEqual(2);
    expect(chapterFilters[0].props.locationId).toEqual(pageId);
    expect(chapterFilters[1].props.locationId).toEqual(chapterId);
    expect(colorFilters.length).toEqual(2);
    expect(colorFilters[0].props.color).toEqual(HighlightColorEnum.Blue);
    expect(colorFilters[1].props.color).toEqual(HighlightColorEnum.Yellow);
  });

  it('dispatches action to print highlights', () => {
    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <Filters />
      </MessageProvider>
    </Provider>);

    const renderedPrintButton = component.root.findByProps({'data-testid': 'hl-print-button'});

    renderer.act(() => {
      renderedPrintButton.props.onClick();
    });

    expect(storeDispatch).toHaveBeenCalledWith(printSummaryHighlights());
  });

  it('removes colors and chapters from filters on click', () => {
    store.dispatch(receiveBook(book));
    store.dispatch(receivePage({...pageInChapter, references: []}));
    store.dispatch(receiveHighlightsTotalCounts({
      'testbook1-testchapter5-uuid': {
        [HighlightColorEnum.Green]: 1,
        [HighlightColorEnum.Blue]: 1,
      },
    }, new Map([[
      'testbook1-testchapter5-uuid',
      assertDefined(findArchiveTreeNode(book.tree, 'testbook1-testchapter5-uuid'), ''),
    ]])));

    dispatch.mockClear();
    storeDispatch.mockClear();

    const component = renderer.create(<Provider store={store}>
      <MessageProvider>
        <Filters />
      </MessageProvider>
    </Provider>);

    let chapterFilters = component.root.findAllByType(FiltersListChapter);
    let colorFilters = component.root.findAllByType(FiltersListColor);

    expect(chapterFilters.length).toEqual(1);
    expect(colorFilters.length).toEqual(2);

    renderer.act(() => {
      chapterFilters[0].findByType(StyledPlainButton).props.onClick();
    });

    expect(storeDispatch).toBeCalledWith(setSummaryFilters({
      locationIds: [],
    }));

    renderer.act(() => {
      colorFilters[0].findByType(StyledPlainButton).props.onClick();
    });

    expect(storeDispatch).toBeCalledWith(setSummaryFilters({
      colors: [ HighlightColorEnum.Green ],
    }));

    chapterFilters = component.root.findAllByType(FiltersListChapter);
    colorFilters = component.root.findAllByType(FiltersListColor);

    expect(chapterFilters.length).toEqual(0);
    expect(colorFilters.length).toEqual(1);
  });
});
