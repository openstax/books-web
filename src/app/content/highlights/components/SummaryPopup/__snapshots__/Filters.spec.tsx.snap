// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters matches snapshot 1`] = `
.c9 {
  display: inline-block;
  vertical-align: -.125em;
  overflow: hidden;
}

.c7 {
  cursor: pointer;
  border: none;
  padding: 0;
  background: none;
}

.c5 {
  cursor: pointer;
}

.c3 > *:not(.c4) {
  -webkit-animation: 100ms bcCCNc ease-out;
  animation: 100ms bcCCNc ease-out;
  position: absolute;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,0.1);
  border: 1px solid #d5d5d5;
  top: calc(100% + 0.4rem);
  left: 0;
}

.c2 {
  overflow: visible;
  position: relative;
}

.c11 {
  cursor: pointer;
  border: none;
  padding: 0;
  background: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #5e6062;
  height: 100%;
  min-width: 45px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11:hover,
.c11:focus {
  outline: none;
  color: #424242;
}

.c13 {
  font-weight: 600;
  font-family: Neue Helvetica W01;
  font-size: 1.6rem;
  line-height: 2.5rem;
  margin: 0 0 0 0.5rem;
}

.c12 {
  height: 2.5rem;
  width: 2.5rem;
  padding: 0.4rem;
}

.c15 {
  cursor: pointer;
  border: none;
  padding: 0;
  background: none;
  padding: 0.5rem;
}

.c15 svg {
  height: 1rem;
  width: 1rem;
}

.c16 {
  max-width: 80px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-transform: capitalize;
}

.c14 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  margin: 0;
  padding: 2rem 0;
  list-style: none;
}

.c14 li {
  margin-right: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 {
  color: #5e6062;
  width: 1rem;
  height: 2rem;
  margin-left: 0.8rem;
}

.c10 {
  min-width: auto;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
  margin-left: auto;
}

.c6 > div {
  outline: none;
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  color: #5e6062;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  overflow: visible;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 2.1rem 3.2rem 0 3.2rem;
  background: #fff;
}

.c0 .c4 {
  font-weight: bold;
}

.c0 .c1:first-of-type {
  margin-right: 8rem;
}

@media screen and (max-width:75em) {
  .c13 {
    display: none;
  }
}

@media screen and (max-width:75em) {
  .c14 {
    display: none;
  }
}

@media screen and (max-width:75em) {
  .c0 {
    padding: 0 1.6rem;
    height: 3.6rem;
  }
}

@media screen and (max-width:75em) {
  .c0 .c1:first-of-type {
    margin-right: 4.8rem;
  }
}

@media print {
  .c0 {
    display: none;
  }
}

<div
  className="c0"
>
  <div
    className="c1 c2 c3"
  >
    <button
      className="c4 c5 c6 c7"
      onClick={[Function]}
    >
      <div
        tabIndex={-1}
      >
        Chapter
        <svg
          aria-hidden="true"
          className="c8 c9"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 320 512"
        >
          <path
            d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
  </div>
  <div
    className="c1 c2 c3"
  >
    <button
      className="c4 c5 c6 c7"
      onClick={[Function]}
    >
      <div
        tabIndex={-1}
      >
        Color
        <svg
          aria-hidden="true"
          className="c8 c9"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 320 512"
        >
          <path
            d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
  </div>
  <button
    aria-label="print"
    className="c10 c11"
    data-testid="print"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="c12 c9"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M448 192V77.25c0-8.49-3.37-16.62-9.37-22.63L393.37 9.37c-6-6-14.14-9.37-22.63-9.37H96C78.33 0 64 14.33 64 32v160c-35.35 0-64 28.65-64 64v112c0 8.84 7.16 16 16 16h48v96c0 17.67 14.33 32 32 32h320c17.67 0 32-14.33 32-32v-96h48c8.84 0 16-7.16 16-16V256c0-35.35-28.65-64-64-64zm-64 256H128v-96h256v96zm0-224H128V64h192v48c0 8.84 7.16 16 16 16h48v96zm48 72c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
        fill="currentColor"
      />
    </svg>
    <span
      className="c13"
    >
      Print
    </span>
  </button>
  <ul
    className="c14"
  >
    <li>
      <button
        className="c15"
        onClick={[Function]}
      >
        <svg
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
      <span
        className="c16"
      >
        yellow
      </span>
    </li>
    <li>
      <button
        className="c15"
        onClick={[Function]}
      >
        <svg
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
      <span
        className="c16"
      >
        green
      </span>
    </li>
    <li>
      <button
        className="c15"
        onClick={[Function]}
      >
        <svg
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
      <span
        className="c16"
      >
        blue
      </span>
    </li>
    <li>
      <button
        className="c15"
        onClick={[Function]}
      >
        <svg
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
      <span
        className="c16"
      >
        purple
      </span>
    </li>
    <li>
      <button
        className="c15"
        onClick={[Function]}
      >
        <svg
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
      <span
        className="c16"
      >
        pink
      </span>
    </li>
  </ul>
</div>
`;
